// @fluentui/react-migration-v8-v9/lib/components/Menu/shimMenuProps.js
import * as React from 'react';
import { Icon } from '@fluentui/react';
export const shimMenuProps = (props)=>{
    return {
        hasIcons: props === null || props === void 0 ? void 0 : props.items.some((i)=>i.iconProps),
        hasCheckmarks: props === null || props === void 0 ? void 0 : props.items.some((i)=>i.canCheck || i.checked)
    };
};
export const shimMenuItemProps = (props)=>{
    //TODO: Handle menuItem.onRenderIcon
    const icon = props.iconProps ? /*#__PURE__*/ React.createElement(Icon, props.iconProps) : undefined;
    return {
        ...props,
        'aria-label': props.ariaLabel,
        children: props.text,
        icon,
        secondaryContent: props.secondaryText
    };
};
export const shimMenuItemCheckboxProps = (props)=>{
    return {
        ...shimMenuItemProps(props),
        name: props.text || 'name',
        value: props.value || 'value'
    };
};
export const shimMenuHeaderProps = (props)=>{
    var _props_sectionProps;
    return {
        children: (_props_sectionProps = props.sectionProps) === null || _props_sectionProps === void 0 ? void 0 : _props_sectionProps.title
    };
};
