// @microsoft/sp-image-helper/lib-commonjs/ThumbnailHealingKillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPassOptimizeFormatToVroomKSActivated = exports.isFixAbsoluteUrlFallbackKSActivated = exports.isAdvancedImageEditorKSActivated = exports.isInlineAdvancedCropperKSActivated = exports.isAdvancedEditVersioningKSActivated = exports.isOptimizeJpegKSActivated = exports.isOptimizePngKSActivated = exports.isPreferNoRedirectUpdateKSActivated = exports.isUrlGlobalReplaceKSActivated = exports.isQuickLinksThumbnailViewKSActivated = exports.isDocVizUrlRewriteKSActivated = exports.isVROOMThumbnailRewriteKSActivated = exports.isCDNRewriteKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
/**
 * Contains predefined kill switches across server and client to disable rewrite to specific
 * service during emergency. These kill switches should not be activated in general but only during
 * emergency situation to mitigate live site for a farm or environment as required.
 * In case a kill switch is activated globally to update guid, ensure to also update the server side guid for same.
 *
 * @internal
 */
/**
 * Do not rewrite URL to CND when true
 */
function isCDNRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c2151f9e-ff03-4fa5-8852-10f75e817d4e'
    /* '03/01/2020', 'WEX Disable CDN Rewrite' */
    );
}
exports.isCDNRewriteKSActivated = isCDNRewriteKSActivated;
/**
 * Do not rewrite URL to VROOM Thumbnail when true
 */
function isVROOMThumbnailRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('318c1916-4594-4509-8ab7-cdb100182007'
    /* '03/01/2020', 'WEX Disable VROOM Thumbnail rewrite' */
    );
}
exports.isVROOMThumbnailRewriteKSActivated = isVROOMThumbnailRewriteKSActivated;
/**
 * Do not rewrite URL to getPreview.ashx when true.
 *
 * @remarks
 * Certain environments like Gallatin supports only DocViz
 */
function isDocVizUrlRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('67d7e188-b33c-4fc2-8862-a072c6d8afb4'
    /* '03/01/2020', 'WEX Disable DocViz URL Rewrite' */
    );
}
exports.isDocVizUrlRewriteKSActivated = isDocVizUrlRewriteKSActivated;
function isQuickLinksThumbnailViewKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f1135e46-8b4b-41c8-9af2-008a1596012d'
    /* '6/14/2022', 'FILA fix quicklinks thumbnail not shown consistently for files' */
    );
}
exports.isQuickLinksThumbnailViewKSActivated = isQuickLinksThumbnailViewKSActivated;
function isUrlGlobalReplaceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('2f268f84-0eee-4f0f-a743-fec73e5e76b2'
    /* '08/24/2022', 'WEX Global replace of forward slash and plus symbol in preview urls' */
    );
}
exports.isUrlGlobalReplaceKSActivated = isUrlGlobalReplaceKSActivated;
function isPreferNoRedirectUpdateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('04e0cf44-13db-4023-bba2-67aab83c48eb'
    /* '04/07/2023', 'Update VROOM params from preferNoRedirect=true to prefer=noRedirect' */
    );
}
exports.isPreferNoRedirectUpdateKSActivated = isPreferNoRedirectUpdateKSActivated;
function isOptimizePngKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a7860782-db85-42fc-b1dc-dfb0abaafc97'
    /* '06/27/2023', 'Apply optimization param for PNG files' */
    );
}
exports.isOptimizePngKSActivated = isOptimizePngKSActivated;
function isOptimizeJpegKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f82624b3-0b57-488e-aa1e-4e164321b4e2'
    /* '06/27/2023', 'Apply optimization param for Jpeg files' */
    );
}
exports.isOptimizeJpegKSActivated = isOptimizeJpegKSActivated;
//cross-project KS
function isAdvancedEditVersioningKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('71c16ef8-a7a3-4620-acae-df29823c6e57'
    /* '09/28/2023', 'thomashorton- Enabled saving over the same file name in advanced editor and cropper' */
    );
}
exports.isAdvancedEditVersioningKSActivated = isAdvancedEditVersioningKSActivated;
//cross-project-ks
function isInlineAdvancedCropperKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('79342c97-223e-4c20-aa6d-120bfec637ca'
    /* '02/24/2023', 'thomashorton- use (OMG) advanced image editor library for inline crop' */
    );
}
exports.isInlineAdvancedCropperKSActivated = isInlineAdvancedCropperKSActivated;
// cross-project KS
function isAdvancedImageEditorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('cf9a0630-222f-4c26-876e-5d880e8a77b9'
    /* '04/14/2023', 'thomashorton- Advanced Image Editor KS, cross project also exists in ImageWebPartKillswitches' */
    );
}
exports.isAdvancedImageEditorKSActivated = isAdvancedImageEditorKSActivated;
function isFixAbsoluteUrlFallbackKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('59c6659e-2144-4409-9891-ab44d5ba332e'
    /* '12/06/2023', 'yidansun - Use 'localtion.href' fallback when absolute url is undefiend */
    );
}
exports.isFixAbsoluteUrlFallbackKSActivated = isFixAbsoluteUrlFallbackKSActivated;
function isPassOptimizeFormatToVroomKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8a1a5f98-061b-4883-a3e6-8e78abb9687a'
    /* '11/14/2023', 'pass "format=webp" as  param' for optimization */
    );
}
exports.isPassOptimizeFormatToVroomKSActivated = isPassOptimizeFormatToVroomKSActivated;
