// @fluentui/fluent2-theme/lib-amd/fluent2ComponentStyles.js
define(["require", "exports", "./componentStyles/BasePicker.styles", "./componentStyles/Breadcrumb.styles", "./componentStyles/Button.styles", "./componentStyles/Callout.styles", "./componentStyles/Checkbox.styles", "./componentStyles/ChoiceGroup.styles", "./componentStyles/ChoiceGroupOption.styles", "./componentStyles/ColorPickerGridStyles.styles", "./componentStyles/CommandBar.styles", "./componentStyles/CommandBarButton.styles", "./componentStyles/ContextualMenu.styles", "./componentStyles/Dialog.styles", "./componentStyles/Dropdown.styles", "./componentStyles/MessageBar.styles", "./componentStyles/Modal.styles", "./componentStyles/Pivot.styles", "./componentStyles/SearchBox.styles", "./componentStyles/Slider.styles", "./componentStyles/SpinButton.styles", "./componentStyles/Spinner.styles", "./componentStyles/TagItem.styles", "./componentStyles/TextField.styles", "./componentStyles/Toggle.styles", "./componentStyles/DetailsRowCheck.styles", "./componentStyles/Check.styles"], function (require, exports, BasePicker_styles_1, Breadcrumb_styles_1, Button_styles_1, Callout_styles_1, Checkbox_styles_1, ChoiceGroup_styles_1, ChoiceGroupOption_styles_1, ColorPickerGridStyles_styles_1, CommandBar_styles_1, CommandBarButton_styles_1, ContextualMenu_styles_1, Dialog_styles_1, Dropdown_styles_1, MessageBar_styles_1, Modal_styles_1, Pivot_styles_1, SearchBox_styles_1, Slider_styles_1, SpinButton_styles_1, Spinner_styles_1, TagItem_styles_1, TextField_styles_1, Toggle_styles_1, DetailsRowCheck_styles_1, Check_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.fluent2ComponentStyles = void 0;
    exports.fluent2ComponentStyles = {
        Breadcrumb: {
            styles: Breadcrumb_styles_1.getBreadcrumbStyles,
        },
        CalloutContent: {
            styles: Callout_styles_1.getCalloutContentStyles,
        },
        Check: {
            styles: Check_styles_1.getCheckStyles,
        },
        Checkbox: {
            styles: Checkbox_styles_1.getCheckboxStyles,
        },
        ChoiceGroup: {
            styles: ChoiceGroup_styles_1.getChoiceGroupStyles,
        },
        ChoiceGroupOption: {
            styles: ChoiceGroupOption_styles_1.getChoiceGroupOptionStyles,
        },
        ColorPickerGridCell: {
            styles: ColorPickerGridStyles_styles_1.getColorPickerGridCellStyles,
        },
        CommandBar: {
            styles: CommandBar_styles_1.getCommandBarStyles,
        },
        CommandBarButton: {
            styles: CommandBarButton_styles_1.getCommandBarButtonStyles,
        },
        CompoundButton: {
            styles: Button_styles_1.getDefaultButtonStyles,
        },
        ContextualMenu: {
            styles: ContextualMenu_styles_1.getContextualMenuStyles,
        },
        DefaultButton: {
            styles: Button_styles_1.getDefaultButtonStyles,
        },
        DetailRowCheck: {
            styles: DetailsRowCheck_styles_1.getDetailsRowCheckStyles,
        },
        Dialog: {
            styles: Dialog_styles_1.getDialogStyles,
        },
        DialogContent: {
            styles: Dialog_styles_1.getDialogContentStyles,
        },
        Dropdown: {
            styles: Dropdown_styles_1.getDropdownStyles,
        },
        IconButton: {
            styles: Button_styles_1.getIconButtonStyles,
        },
        MessageBar: {
            styles: MessageBar_styles_1.getMessageBarStyles,
        },
        Modal: {
            styles: Modal_styles_1.getModalStyles,
        },
        Pivot: {
            styles: Pivot_styles_1.getPivotStyles,
        },
        // People Pickers
        NormalPeoplePicker: {
            styles: BasePicker_styles_1.getBasePickerStyles,
        },
        CompactPeoplePicker: {
            styles: BasePicker_styles_1.getBasePickerStyles,
        },
        ListPeoplePickerBase: {
            styles: BasePicker_styles_1.getBasePickerStyles,
        },
        SearchBox: {
            styles: SearchBox_styles_1.getSearchBoxStyles,
        },
        Slider: {
            styles: Slider_styles_1.getSliderStyles,
        },
        SpinButton: {
            styles: SpinButton_styles_1.getSpinButtonStyles,
        },
        Spinner: {
            styles: Spinner_styles_1.getSpinnerStyles,
        },
        TagItem: {
            styles: TagItem_styles_1.getTagItemStyles,
        },
        TagPicker: {
            styles: BasePicker_styles_1.getBasePickerStyles,
        },
        Tag: {
            styles: TagItem_styles_1.getTagItemStyles,
        },
        TextField: {
            styles: TextField_styles_1.getTextFieldStyles,
        },
        Toggle: {
            styles: Toggle_styles_1.getToggleStyles,
        },
    };
});
