// @fluentui/react-migration-v8-v9/lib-commonjs/components/Stack/StackItemShim.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useFlexAlignSelfStyles: function() {
        return useFlexAlignSelfStyles;
    },
    useFlexGrowStyles: function() {
        return useFlexGrowStyles;
    },
    useFlexOrderStyles: function() {
        return useFlexOrderStyles;
    },
    useFlexShrinkStyles: function() {
        return useFlexShrinkStyles;
    },
    useStackItemShimStyles: function() {
        return useStackItemShimStyles;
    }
});
const _react = require("@griffel/react");
const useStackItemShimStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        Bqenvij: "f11ysow2",
        a9b677: "f14z66ap"
    },
    disableShrink: {
        Bnnss6s: "fi64zpg"
    },
    verticalFill: {
        Bqenvij: "f1l02sjl"
    }
}, {
    d: [
        ".f11ysow2{height:auto;}",
        ".f14z66ap{width:auto;}",
        ".fi64zpg{flex-shrink:0;}",
        ".f1l02sjl{height:100%;}"
    ]
});
const useFlexAlignSelfStyles = /*#__PURE__*/ (0, _react.__styles)({
    auto: {
        qb2dma: "fio29c6"
    },
    baseline: {
        qb2dma: "f18ssl25"
    },
    center: {
        qb2dma: "f7nlbp4"
    },
    start: {
        qb2dma: "fjgzulp"
    },
    end: {
        qb2dma: "fpydfdc"
    },
    stretch: {
        qb2dma: "f1ub7u0d"
    }
}, {
    d: [
        ".fio29c6{align-self:auto;}",
        ".f18ssl25{align-self:baseline;}",
        ".f7nlbp4{align-self:center;}",
        ".fjgzulp{align-self:flex-start;}",
        ".fpydfdc{align-self:flex-end;}",
        ".f1ub7u0d{align-self:stretch;}"
    ]
});
const useFlexGrowStyles = /*#__PURE__*/ (0, _react.__styles)({
    "1": {
        Bh6795r: "fqerorx"
    },
    "2": {
        Bh6795r: "fv68sux"
    },
    "3": {
        Bh6795r: "f1g5s877"
    },
    inherit: {
        Bh6795r: "fbrulcr"
    },
    initial: {
        Bh6795r: "fqq5sy3"
    },
    revert: {
        Bh6795r: "f1kc8nlq"
    },
    unset: {
        Bh6795r: "fepjonw"
    }
}, {
    d: [
        ".fqerorx{flex-grow:1;}",
        ".fv68sux{flex-grow:2;}",
        ".f1g5s877{flex-grow:3;}",
        ".fbrulcr{flex-grow:inherit;}",
        ".fqq5sy3{flex-grow:initial;}",
        ".f1kc8nlq{flex-grow:revert;}",
        ".fepjonw{flex-grow:unset;}"
    ]
});
const useFlexShrinkStyles = /*#__PURE__*/ (0, _react.__styles)({
    "0": {
        Bnnss6s: "fi64zpg"
    },
    "1": {
        Bnnss6s: "f1neuvcm"
    },
    "2": {
        Bnnss6s: "f1nje5c6"
    },
    inherit: {
        Bnnss6s: "f11ln2im"
    },
    initial: {
        Bnnss6s: "f187621l"
    },
    revert: {
        Bnnss6s: "f1jq7m8u"
    },
    unset: {
        Bnnss6s: "f10s6wei"
    }
}, {
    d: [
        ".fi64zpg{flex-shrink:0;}",
        ".f1neuvcm{flex-shrink:1;}",
        ".f1nje5c6{flex-shrink:2;}",
        ".f11ln2im{flex-shrink:inherit;}",
        ".f187621l{flex-shrink:initial;}",
        ".f1jq7m8u{flex-shrink:revert;}",
        ".f10s6wei{flex-shrink:unset;}"
    ]
});
const useFlexOrderStyles = /*#__PURE__*/ (0, _react.__styles)({
    "0": {
        Bxn5svx: "f1ftwmyh"
    },
    "1": {
        Bxn5svx: "f15wwbgh"
    },
    "2": {
        Bxn5svx: "f15ak53d"
    },
    "3": {
        Bxn5svx: "f19x901k"
    },
    inherit: {
        Bxn5svx: "ficp5m6"
    },
    initial: {
        Bxn5svx: "f1xfv85z"
    },
    unset: {
        Bxn5svx: "f1ex5c7r"
    },
    revert: {
        Bxn5svx: "fgu8tny"
    },
    "-3": {
        Bxn5svx: "folhkm8"
    },
    "-2": {
        Bxn5svx: "f1yml5i9"
    },
    "-1": {
        Bxn5svx: "f1l5m61u"
    }
}, {
    d: [
        ".f1ftwmyh{order:0;}",
        ".f15wwbgh{order:1;}",
        ".f15ak53d{order:2;}",
        ".f19x901k{order:3;}",
        ".ficp5m6{order:inherit;}",
        ".f1xfv85z{order:initial;}",
        ".f1ex5c7r{order:unset;}",
        ".fgu8tny{order:revert;}",
        ".folhkm8{order:-3;}",
        ".f1yml5i9{order:-2;}",
        ".f1l5m61u{order:-1;}"
    ]
}); //# sourceMappingURL=StackItemShim.styles.js.map
