// adaptivecards-designer/lib/open-json-schema-dialog.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenJsonSchemaDialog = void 0;
var Adaptive = require("adaptivecards");
var dialog_1 = require("./dialog");
var adaptivecards_templating_1 = require("adaptivecards-templating");
var OpenJsonSchemaDialog = /** @class */ (function (_super) {
    __extends(OpenJsonSchemaDialog, _super);
    function OpenJsonSchemaDialog() {
        var _this = _super.call(this) || this;
        _this.submitButton = new dialog_1.DialogButton("Create");
        _this.submitButton.onClick = function (sender) {
            _this.close();
        };
        _this.buttons = [_this.submitButton];
        return _this;
    }
    OpenJsonSchemaDialog.prototype.open = function () {
        _super.prototype.open.call(this);
        this._editor.layout();
    };
    OpenJsonSchemaDialog.prototype.renderPreview = function () {
        var _a;
        var renderedElement = document.createElement("div");
        renderedElement.className = "acd-open-code-dialog-preview";
        var card = new Adaptive.AdaptiveCard();
        if ((_a = this.output) === null || _a === void 0 ? void 0 : _a.cardPayload) {
            try {
                var cardData = JSON.parse(this.output.cardPayload);
                card.parse(cardData);
            }
            catch (e) {
                console.error(e);
            }
        }
        else {
            card.parse({
                "type": "AdaptiveCard",
                "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
                "version": "1.0",
                "body": [
                    {
                        "type": "TextBlock",
                        "text": "Input JSON Schema to create a card!",
                        "wrap": true,
                        "size": "Large"
                    }
                ]
            });
        }
        card.render();
        card.renderedElement.style.width = "100%";
        renderedElement.appendChild(card.renderedElement);
        return renderedElement;
    };
    OpenJsonSchemaDialog.prototype.refreshPreview = function () {
        this._previewContainer.removeChild(this._previewElement);
        this._previewElement = this.renderPreview();
        this._previewContainer.appendChild(this._previewElement);
    };
    OpenJsonSchemaDialog.prototype.renderEditor = function () {
        var _this = this;
        var renderedElement = document.createElement("div");
        renderedElement.className = "acd-open-code-dialog-editor";
        var monaco = window["monaco"];
        this._editor = monaco.editor.create(renderedElement, {
            folding: true,
            fontSize: 13.5,
            language: 'json',
        });
        this._editor.layout();
        this._editor.onDidChangeModelContent(function () {
            try {
                var model = _this._editor.getValue();
                var json = JSON.parse(model);
                var card = (0, adaptivecards_templating_1.JSONSchemaCard)(json);
                _this._output = {
                    cardPayload: JSON.stringify(card),
                };
                _this.refreshPreview();
            }
            catch (e) {
                console.error(e);
            }
        });
        return renderedElement;
    };
    OpenJsonSchemaDialog.prototype.renderContent = function () {
        var renderedElement = document.createElement("div");
        renderedElement.className = "acd-open-code-dialog-container";
        var editorContainer = document.createElement("div");
        editorContainer.className = "acd-open-code-dialog-editor-container";
        renderedElement.appendChild(editorContainer);
        var editorLabel = document.createElement("h3");
        editorLabel.innerText = "JSON Schema";
        editorContainer.appendChild(editorLabel);
        var editorElement = this.renderEditor();
        editorContainer.appendChild(editorElement);
        this._previewContainer = document.createElement("div");
        this._previewContainer.className = "acd-open-code-dialog-preview-container";
        renderedElement.appendChild(this._previewContainer);
        this._previewElement = this.renderPreview();
        this._previewContainer.appendChild(this._previewElement);
        return renderedElement;
    };
    Object.defineProperty(OpenJsonSchemaDialog.prototype, "output", {
        get: function () {
            return this._output;
        },
        enumerable: false,
        configurable: true
    });
    return OpenJsonSchemaDialog;
}(dialog_1.Dialog));
exports.OpenJsonSchemaDialog = OpenJsonSchemaDialog;
