// adaptivecards-designer/lib/open-sample-dialog.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenSampleDialog = void 0;
var ACData = require("adaptivecards-templating");
var Adaptive = require("adaptivecards");
var dialog_1 = require("./dialog");
var adaptivecards_controls_1 = require("adaptivecards-controls");
var OpenSampleItem = /** @class */ (function () {
    function OpenSampleItem(props) {
        this.props = props;
    }
    OpenSampleItem.getNewItemId = function (prefix) {
        var newId = prefix + "-" + OpenSampleItem._id;
        OpenSampleItem._id++;
        return newId;
    };
    OpenSampleItem.prototype.render = function () {
        var _this = this;
        var _a;
        var newItemId = OpenSampleItem.getNewItemId("acd-open-sample-item-title");
        var element = document.createElement("div");
        element.className = "acd-open-sample-item";
        element.tabIndex = 0;
        element.setAttribute("aria-labelledBy", newItemId);
        element.setAttribute("role", "listitem");
        element.onclick = (_a = this.props.onClick) !== null && _a !== void 0 ? _a : (function (e) {
            _this.onCardSelected();
        });
        element.onkeyup = function (e) {
            switch (e.key) {
                case adaptivecards_controls_1.Constants.keys.enter:
                    // If onKeyEnterEvent has been specified, we will follow that logic for selecting an OpenSampleItem
                    if (_this.props.onKeyEnterEvent) {
                        _this.props.onKeyEnterEvent(e);
                    }
                    else {
                        // If onKeyEnterEvent is not specified, we assume OpenSampleItem should be selected and loaded as usual
                        _this.onCardSelected();
                    }
                    break;
                case adaptivecards_controls_1.Constants.keys.down:
                    _this.navigateToNextElement(element);
                    break;
                case adaptivecards_controls_1.Constants.keys.up:
                    _this.navigateToPreviousElement(element);
                    break;
                case adaptivecards_controls_1.Constants.keys.right:
                    _this.navigateToNextElement(element);
                    break;
                case adaptivecards_controls_1.Constants.keys.left:
                    _this.navigateToPreviousElement(element);
                    break;
                default:
                    break;
            }
        };
        var thumbnailHost = document.createElement("div");
        thumbnailHost.className = "acd-open-sample-item-thumbnail";
        if (this.props.cardData instanceof Function) {
            var spinner_1 = document.createElement("div");
            spinner_1.className = "acd-spinner";
            spinner_1.style.width = "28px";
            spinner_1.style.height = "28px";
            thumbnailHost.appendChild(spinner_1);
            var cardData = this.props.cardData(function (cardData) {
                thumbnailHost.removeChild(spinner_1);
                if (cardData.thumbnail instanceof Function) {
                    thumbnailHost.appendChild(cardData.thumbnail());
                }
                else if (cardData.thumbnail) {
                    thumbnailHost.appendChild(cardData.thumbnail);
                }
            });
            if (cardData) {
                thumbnailHost.removeChild(spinner_1);
                if (cardData.thumbnail instanceof Function) {
                    thumbnailHost.appendChild(cardData.thumbnail());
                }
                else if (cardData.thumbnail) {
                    thumbnailHost.appendChild(cardData.thumbnail);
                }
            }
        }
        else if (this.props.cardData) {
            if (this.props.cardData.thumbnail instanceof Function) {
                thumbnailHost.appendChild(this.props.cardData.thumbnail());
            }
            else if (this.props.cardData.thumbnail) {
                thumbnailHost.appendChild(this.props.cardData.thumbnail);
            }
        }
        var displayNameElement = document.createElement("div");
        displayNameElement.className = "acd-open-sample-item-title";
        displayNameElement.id = newItemId;
        displayNameElement.innerText = this.props.label;
        element.appendChild(thumbnailHost);
        element.appendChild(displayNameElement);
        return element;
    };
    OpenSampleItem.prototype.navigateToNextElement = function (element) {
        if (element.nextSibling) {
            element.nextSibling.focus();
        }
        else {
            if (this.onNavigateToNextList) {
                this.onNavigateToNextList();
            }
        }
    };
    OpenSampleItem.prototype.navigateToPreviousElement = function (element) {
        if (element.previousSibling) {
            element.previousSibling.focus();
        }
        else {
            if (this.onNavigateToPreviousList) {
                this.onNavigateToPreviousList();
            }
        }
    };
    OpenSampleItem.prototype.onCardSelected = function () {
        if (this.onComplete) {
            if (this.props.cardData instanceof Function) {
                var cardData = this.props.cardData(this.onComplete);
                if (cardData) {
                    this.onComplete(cardData);
                }
            }
            else if (this.props.cardData) {
                this.onComplete(this.props.cardData);
            }
        }
    };
    OpenSampleItem._id = 0;
    return OpenSampleItem;
}());
var OpenSampleDialog = /** @class */ (function (_super) {
    __extends(OpenSampleDialog, _super);
    function OpenSampleDialog(props) {
        var _this = _super.call(this) || this;
        _this.props = props;
        _this._listElements = [];
        return _this;
    }
    OpenSampleDialog.prototype.renderSection = function (title, items) {
        var _this = this;
        var renderedElement = document.createElement("div");
        if (title) {
            var titleElement = document.createElement("div");
            titleElement.className = "acd-dialog-title";
            titleElement.innerText = title;
            titleElement.style.marginTop = "10px";
            renderedElement.appendChild(titleElement);
        }
        var listElement = document.createElement("div");
        listElement.className = "acd-open-sample-item-container";
        listElement.setAttribute("role", "list");
        this._listElements.push(listElement);
        for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
            var item = items_1[_i];
            if (!item)
                continue;
            var itemElement = new OpenSampleItem(item);
            itemElement.onComplete = function (output) {
                _this._output = output;
                _this.close();
            };
            itemElement.onNavigateToNextList = function () {
                var _a;
                // Find the index of the current list
                var currentIndex = _this._listElements.indexOf(listElement);
                // If the next index has a valid list, we want to navigate to the first element
                if ((currentIndex !== -1) && (currentIndex + 1 < _this._listElements.length) && ((_a = _this._listElements.at(currentIndex + 1)) === null || _a === void 0 ? void 0 : _a.firstChild)) {
                    _this._listElements.at(currentIndex + 1).firstChild.focus();
                }
            };
            itemElement.onNavigateToPreviousList = function () {
                var _a;
                // Find the index of the current list
                var currentIndex = _this._listElements.indexOf(listElement);
                // If the previous index has a valid list, we want to navigate to the last element
                if ((currentIndex !== -1) && (currentIndex - 1 >= 0) && ((_a = _this._listElements.at(currentIndex - 1)) === null || _a === void 0 ? void 0 : _a.lastChild)) {
                    _this._listElements.at(currentIndex - 1).lastChild.focus();
                }
            };
            listElement.appendChild(itemElement.render());
        }
        renderedElement.appendChild(listElement);
        return renderedElement;
    };
    OpenSampleDialog.prototype.renderContent = function () {
        var _this = this;
        var renderedElement = document.createElement("div");
        renderedElement.style.overflow = "auto";
        var featuredSection = this.renderSection(null, OpenSampleDialog._builtinItems.concat(this.props.handlers));
        renderedElement.appendChild(featuredSection);
        if (this.props.catalogue) {
            var divider = document.createElement("hr");
            renderedElement.appendChild(divider);
            this.props.catalogue.onDownloaded = function (sender) {
                if (sender.isDownloaded) {
                    var catalogueSection = _this.renderSection("Explore", _this.props.catalogue.entries.map(function (entry) {
                        return {
                            label: entry.displayName,
                            cardData: function (callback) {
                                entry.onDownloaded = function (sender) {
                                    var success = sender.cardPayloadDownloaded;
                                    if (success) {
                                        try {
                                            var cardPayload = JSON.parse(sender.cardPayload);
                                            var card = new Adaptive.AdaptiveCard();
                                            var cardData = sender.sampleData ?
                                                new ACData.Template(cardPayload).expand({
                                                    $root: JSON.parse(sender.sampleData)
                                                })
                                                : cardPayload;
                                            card.parse(cardData);
                                            card.render();
                                            card.renderedElement.style.width = "100%";
                                            return callback({
                                                cardPayload: entry.cardPayload,
                                                sampleData: entry.sampleData,
                                                sampleHostData: "{}",
                                                thumbnail: card.renderedElement,
                                            });
                                        }
                                        catch (e) {
                                            // Swallow the exception
                                            console.error("Unable to load card sample. Error: " + e);
                                            success = false;
                                        }
                                    }
                                };
                                entry.download();
                            },
                        };
                    }));
                    renderedElement.appendChild(catalogueSection);
                }
                else {
                    console.error("Sender is not downloaded");
                }
            };
            this.props.catalogue.download();
        }
        var firstChild = featuredSection.firstElementChild;
        firstChild.focus();
        return renderedElement;
    };
    Object.defineProperty(OpenSampleDialog.prototype, "output", {
        get: function () {
            return this._output;
        },
        enumerable: false,
        configurable: true
    });
    OpenSampleDialog._builtinItems = [
        {
            label: "Blank Card",
            cardData: {
                cardPayload: JSON.stringify({
                    type: "AdaptiveCard",
                    $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
                    version: "1.0",
                    body: []
                }),
            },
        },
    ];
    return OpenSampleDialog;
}(dialog_1.Dialog));
exports.OpenSampleDialog = OpenSampleDialog;
