// adaptivecards-designer/lib/strings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strings = void 0;
var containers_1 = require("./containers");
var Strings = /** @class */ (function () {
    function Strings() {
    }
    Strings.toolboxes = {
        toolPalette: {
            title: "CARD ELEMENTS"
        },
        cardEditor: {
            title: "CARD PAYLOAD EDITOR"
        },
        sampleDataEditor: {
            title: "SAMPLE DATA EDITOR"
        },
        sampleHostDataEditor: {
            title: "SAMPLE HOST DATA EDITOR"
        },
        propertySheet: {
            title: "ELEMENT PROPERTIES",
            commands: {
                bindData: {
                    displayText: function () { return "..."; },
                    accessibleText: function (propertyLabel) { return "Data Binding"; }
                }
            }
        },
        cardStructure: {
            title: "CARD STRUCTURE"
        },
        dataStructure: {
            title: "DATA STRUCTURE"
        }
    };
    Strings.loadingEditor = "Loading editor...";
    Strings.widgetOverflowWarning = "[Designer] We have adjusted your card to hide overflow elements. Please ensure all card elements are contained within the proper card height per Widgets Board design guidelines. Only Carousel pagination dots can be present in the bottom " + containers_1.WidgetContainer.widgetPadding + "px margin of a widget.";
    return Strings;
}());
exports.Strings = Strings;
