// @fluentui/fluent2-theme/lib-commonjs/componentStyles/SearchBox.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSearchBoxStyles = void 0;
var inputStyleHelpers_utils_1 = require("./inputStyleHelpers.utils");
function getSearchBoxStyles(props) {
    var hasFocus = props.hasFocus, underlined = props.underlined, disabled = props.disabled, theme = props.theme;
    var palette = theme.palette, semanticColors = theme.semanticColors;
    var restBottomBorder = "1px solid ".concat(palette.neutralPrimary);
    var styles = {
        root: [
            {
                borderRadius: underlined ? 0 : theme.effects.roundedCorner4,
                borderBottom: restBottomBorder,
            },
            hasFocus && [
                'is-active',
                {
                    position: 'relative',
                    borderColor: 'unset',
                },
                (0, inputStyleHelpers_utils_1.getFluent2InputFocusStyles)(theme),
            ],
            disabled && [
                'is-disabled',
                (0, inputStyleHelpers_utils_1.getFluent2InputDisabledStyles)(theme),
                { borderBottom: "1px solid ".concat(semanticColors.disabledBorder) },
            ],
        ],
        field: [
            disabled && {
                '::placeholder': {
                    color: semanticColors.disabledText,
                },
            },
        ],
        iconContainer: [
            {
                color: theme.palette.neutralPrimary,
            },
            disabled && [
                'is-disabled',
                {
                    color: semanticColors.disabledText,
                },
            ],
        ],
    };
    return styles;
}
exports.getSearchBoxStyles = getSearchBoxStyles;
