// @fluentui/fluent2-theme/lib-amd/componentStyles/Dialog.styles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getDialogContentStyles = exports.getDialogStyles = void 0;
    function getDialogStyles(props) {
        var theme = props.theme;
        var effects = theme.effects;
        var styles = {
            main: {
                borderRadius: effects.roundedCorner8,
                boxShadow: effects.elevation64,
            },
        };
        return styles;
    }
    exports.getDialogStyles = getDialogStyles;
    function getDialogContentStyles(props) {
        var theme = props.theme;
        var effects = theme.effects;
        var styles = {
            topButton: {
                '&:hover': {
                    '.ms-Dialog-button': {
                        borderRadius: effects.roundedCorner4,
                    },
                },
            },
            button: {
                borderRadius: effects.roundedCorner4,
            },
        };
        return styles;
    }
    exports.getDialogContentStyles = getDialogContentStyles;
});
