// @pnp/graph/list-item/types.js
import { __decorate } from "tslib";
import { _GraphCollection, graphInvokableFactory, _GraphInstance, GraphCollection, graphPost } from "../graphqueryable.js";
import { defaultPath, deleteable, updateable, getById, addable } from "../decorators.js";
/**
 * Represents a list item entity
 */
let _ListItem = class _ListItem extends _GraphInstance {
    /**
     * Method for retrieving the versions of a list item.
     * @returns IListItemVersion
     */
    get versions() {
        return GraphCollection(this, "versions");
    }
};
_ListItem = __decorate([
    deleteable(),
    updateable()
], _ListItem);
export { _ListItem };
export const ListItem = graphInvokableFactory(_ListItem);
/**
 * Describes a collection of list item objects
 *
 */
let _ListItems = class _ListItems extends _GraphCollection {
};
_ListItems = __decorate([
    defaultPath("items"),
    getById(ListItem),
    addable()
], _ListItems);
export { _ListItems };
export const ListItems = graphInvokableFactory(_ListItems);
/**
 * Represents a document set version
 */
let _DocumentSetVersion = class _DocumentSetVersion extends _GraphInstance {
    /**
     * Restore a document set version
     *
     */
    async restore() {
        return graphPost(DocumentSetVersion(this, "restore"));
    }
};
_DocumentSetVersion = __decorate([
    deleteable()
], _DocumentSetVersion);
export { _DocumentSetVersion };
export const DocumentSetVersion = graphInvokableFactory(_DocumentSetVersion);
/**
 * Describes a collection of document set versions
 *
 */
let _DocumentSetVersions = class _DocumentSetVersions extends _GraphCollection {
};
_DocumentSetVersions = __decorate([
    defaultPath("documentSetVersions"),
    getById(DocumentSetVersion),
    addable()
], _DocumentSetVersions);
export { _DocumentSetVersions };
export const DocumentSetVersions = graphInvokableFactory(_DocumentSetVersions);
