// @fluentui/fluent2-theme/lib-commonjs/fluent2ComponentStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fluent2ComponentStyles = void 0;
var BasePicker_styles_1 = require("./componentStyles/BasePicker.styles");
var Breadcrumb_styles_1 = require("./componentStyles/Breadcrumb.styles");
var Button_styles_1 = require("./componentStyles/Button.styles");
var Callout_styles_1 = require("./componentStyles/Callout.styles");
var Checkbox_styles_1 = require("./componentStyles/Checkbox.styles");
var ChoiceGroup_styles_1 = require("./componentStyles/ChoiceGroup.styles");
var ChoiceGroupOption_styles_1 = require("./componentStyles/ChoiceGroupOption.styles");
var ColorPickerGridStyles_styles_1 = require("./componentStyles/ColorPickerGridStyles.styles");
var CommandBar_styles_1 = require("./componentStyles/CommandBar.styles");
var CommandBarButton_styles_1 = require("./componentStyles/CommandBarButton.styles");
var ContextualMenu_styles_1 = require("./componentStyles/ContextualMenu.styles");
var Dialog_styles_1 = require("./componentStyles/Dialog.styles");
var Dropdown_styles_1 = require("./componentStyles/Dropdown.styles");
var MessageBar_styles_1 = require("./componentStyles/MessageBar.styles");
var Modal_styles_1 = require("./componentStyles/Modal.styles");
var Pivot_styles_1 = require("./componentStyles/Pivot.styles");
var SearchBox_styles_1 = require("./componentStyles/SearchBox.styles");
var Slider_styles_1 = require("./componentStyles/Slider.styles");
var SpinButton_styles_1 = require("./componentStyles/SpinButton.styles");
var Spinner_styles_1 = require("./componentStyles/Spinner.styles");
var TagItem_styles_1 = require("./componentStyles/TagItem.styles");
var TextField_styles_1 = require("./componentStyles/TextField.styles");
var Toggle_styles_1 = require("./componentStyles/Toggle.styles");
var DetailsRowCheck_styles_1 = require("./componentStyles/DetailsRowCheck.styles");
var Check_styles_1 = require("./componentStyles/Check.styles");
exports.fluent2ComponentStyles = {
    Breadcrumb: {
        styles: Breadcrumb_styles_1.getBreadcrumbStyles,
    },
    CalloutContent: {
        styles: Callout_styles_1.getCalloutContentStyles,
    },
    Check: {
        styles: Check_styles_1.getCheckStyles,
    },
    Checkbox: {
        styles: Checkbox_styles_1.getCheckboxStyles,
    },
    ChoiceGroup: {
        styles: ChoiceGroup_styles_1.getChoiceGroupStyles,
    },
    ChoiceGroupOption: {
        styles: ChoiceGroupOption_styles_1.getChoiceGroupOptionStyles,
    },
    ColorPickerGridCell: {
        styles: ColorPickerGridStyles_styles_1.getColorPickerGridCellStyles,
    },
    CommandBar: {
        styles: CommandBar_styles_1.getCommandBarStyles,
    },
    CommandBarButton: {
        styles: CommandBarButton_styles_1.getCommandBarButtonStyles,
    },
    CompoundButton: {
        styles: Button_styles_1.getDefaultButtonStyles,
    },
    ContextualMenu: {
        styles: ContextualMenu_styles_1.getContextualMenuStyles,
    },
    DefaultButton: {
        styles: Button_styles_1.getDefaultButtonStyles,
    },
    DetailRowCheck: {
        styles: DetailsRowCheck_styles_1.getDetailsRowCheckStyles,
    },
    Dialog: {
        styles: Dialog_styles_1.getDialogStyles,
    },
    DialogContent: {
        styles: Dialog_styles_1.getDialogContentStyles,
    },
    Dropdown: {
        styles: Dropdown_styles_1.getDropdownStyles,
    },
    IconButton: {
        styles: Button_styles_1.getIconButtonStyles,
    },
    MessageBar: {
        styles: MessageBar_styles_1.getMessageBarStyles,
    },
    Modal: {
        styles: Modal_styles_1.getModalStyles,
    },
    Pivot: {
        styles: Pivot_styles_1.getPivotStyles,
    },
    // People Pickers
    NormalPeoplePicker: {
        styles: BasePicker_styles_1.getBasePickerStyles,
    },
    CompactPeoplePicker: {
        styles: BasePicker_styles_1.getBasePickerStyles,
    },
    ListPeoplePickerBase: {
        styles: BasePicker_styles_1.getBasePickerStyles,
    },
    SearchBox: {
        styles: SearchBox_styles_1.getSearchBoxStyles,
    },
    Slider: {
        styles: Slider_styles_1.getSliderStyles,
    },
    SpinButton: {
        styles: SpinButton_styles_1.getSpinButtonStyles,
    },
    Spinner: {
        styles: Spinner_styles_1.getSpinnerStyles,
    },
    TagItem: {
        styles: TagItem_styles_1.getTagItemStyles,
    },
    TagPicker: {
        styles: BasePicker_styles_1.getBasePickerStyles,
    },
    Tag: {
        styles: TagItem_styles_1.getTagItemStyles,
    },
    TextField: {
        styles: TextField_styles_1.getTextFieldStyles,
    },
    Toggle: {
        styles: Toggle_styles_1.getToggleStyles,
    },
};
