// @fluentui/react-migration-v8-v9/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ActionButtonShim: function() {
        return _index.ActionButtonShim;
    },
    ButtonShim: function() {
        return _index.ButtonShim;
    },
    CheckboxShim: function() {
        return _index3.CheckboxShim;
    },
    CommandButtonShim: function() {
        return _index.CommandButtonShim;
    },
    CompoundButtonShim: function() {
        return _index.CompoundButtonShim;
    },
    DefaultButtonShim: function() {
        return _index.DefaultButtonShim;
    },
    MenuButtonShim: function() {
        return _index.MenuButtonShim;
    },
    MenuItemShim: function() {
        return _index1.MenuItemShim;
    },
    PrimaryButtonShim: function() {
        return _index.PrimaryButtonShim;
    },
    StackItemShim: function() {
        return _index2.StackItemShim;
    },
    StackShim: function() {
        return _index2.StackShim;
    },
    ToggleButtonShim: function() {
        return _index.ToggleButtonShim;
    },
    black: function() {
        return _index4.black;
    },
    blackAlpha: function() {
        return _index4.blackAlpha;
    },
    brandWeb: function() {
        return _index4.brandWeb;
    },
    createBrandVariants: function() {
        return _index4.createBrandVariants;
    },
    createV8Theme: function() {
        return _index4.createV8Theme;
    },
    createV9Theme: function() {
        return _index4.createV9Theme;
    },
    grey: function() {
        return _index4.grey;
    },
    grey14Alpha: function() {
        return _index4.grey14Alpha;
    },
    hcButtonFace: function() {
        return _index4.hcButtonFace;
    },
    hcButtonText: function() {
        return _index4.hcButtonText;
    },
    hcCanvas: function() {
        return _index4.hcCanvas;
    },
    hcCanvasText: function() {
        return _index4.hcCanvasText;
    },
    hcDisabled: function() {
        return _index4.hcDisabled;
    },
    hcHighlight: function() {
        return _index4.hcHighlight;
    },
    hcHighlightText: function() {
        return _index4.hcHighlightText;
    },
    hcHyperlink: function() {
        return _index4.hcHyperlink;
    },
    sharedColors: function() {
        return _index4.sharedColors;
    },
    shimButtonProps: function() {
        return _index.shimButtonProps;
    },
    shimMenuProps: function() {
        return _index1.shimMenuProps;
    },
    white: function() {
        return _index4.white;
    },
    whiteAlpha: function() {
        return _index4.whiteAlpha;
    }
});
const _index = require("./components/Button/index");
const _index1 = require("./components/Menu/index");
const _index2 = require("./components/Stack/index");
const _index3 = require("./components/Checkbox/index");
const _index4 = require("./components/Theme/index");
