// @pnp/graph/sites/types.js
import { __decorate } from "tslib";
import { combine } from "@pnp/core";
import { defaultPath } from "../decorators.js";
import { graphInvokableFactory, _GraphCollection, _GraphInstance, graphPost } from "../graphqueryable.js";
import { body } from "@pnp/queryable";
/**
 * Sites
 */
let _Sites = class _Sites extends _GraphCollection {
    /**
     * Gets the team site for the group
     */
    get root() {
        return Site(this, "root");
    }
    getById(id) {
        return Site(this, id);
    }
    /**
     * Get a Site by URL
     * @param hostname: string, the host of the site e.g. "contoso.sharepoint.com"
     * @param siteUrl: string, the server relative url of the site e.g. "/sites/teamsite1"
     * @returns ISite
    */
    async getByUrl(hostname, siteUrl) {
        return Site(this, `${hostname}:${combine("/", siteUrl)}:`).rebase();
    }
    /**
     * List sites across geographies in an organization. This API can also be used to enumerate all sites in a non-multi-geo tenant.
     *
     * @returns A ISites collection which can be used with async iteration to page through the collection
     */
    getAllSites() {
        return Sites(this, "getAllSites");
    }
};
_Sites = __decorate([
    defaultPath("sites")
], _Sites);
export { _Sites };
export const Sites = graphInvokableFactory(_Sites);
/**
 * Site
 */
export class _Site extends _GraphInstance {
    get sites() {
        return Sites(this);
    }
    /**
     * Rebases this ISite instances to ensure it is of the pattern /sites/{site id} regardless of how it was first retrieved
     */
    async rebase() {
        const siteInfo = await Site(this).select("id")();
        return Site([this, `sites/${siteInfo.id}`]);
    }
}
export const Site = graphInvokableFactory(_Site);
/**
 * Followed Sites
 *
 * Note: At this time listing a user's followed sites is not supported with app-only permissions
 */
let _FollowedSites = class _FollowedSites extends _GraphCollection {
    /**
     * Adds site(s) to the user's collection of followed sites
     *
     * @param siteIds The collection of site ids to add
     * @returns Site info for the newly followed sites
     */
    add(...siteIds) {
        return graphPost(FollowedSites(this, "add"), body({ value: siteIds.map(id => ({ id })) }));
    }
    /**
     * REmoves site(s) to the user's collection of followed sites
     *
     * @param siteIds The collection of site ids to remove
     */
    remove(...siteIds) {
        return graphPost(FollowedSites(this, "remove"), body({ value: siteIds.map(id => ({ id })) }));
    }
};
_FollowedSites = __decorate([
    defaultPath("followedsites")
], _FollowedSites);
export { _FollowedSites };
export const FollowedSites = graphInvokableFactory(_FollowedSites);
