// @pnp/graph/admin/people.js
import { __decorate } from "tslib";
import { graphInvokableFactory, _GraphInstance, _GraphCollection } from "../graphqueryable.js";
import { defaultPath, updateable, addable, getById, deleteable } from "../decorators.js";
let _PeopleAdmin = class _PeopleAdmin extends _GraphInstance {
    get profileCardProperties() {
        return ProfileCardProperties(this);
    }
    get pronounSettings() {
        return PronounSettings(this);
    }
};
_PeopleAdmin = __decorate([
    defaultPath("people")
], _PeopleAdmin);
export { _PeopleAdmin };
export const PeopleAdmin = graphInvokableFactory(_PeopleAdmin);
/**
* People Pronoun Settings
*/
let _PronounSettings = class _PronounSettings extends _GraphInstance {
};
_PronounSettings = __decorate([
    defaultPath("pronouns"),
    updateable()
], _PronounSettings);
export { _PronounSettings };
export const PronounSettings = graphInvokableFactory(_PronounSettings);
/**
* Profilecard Property
*/
let _ProfileCardProperty = class _ProfileCardProperty extends _GraphInstance {
};
_ProfileCardProperty = __decorate([
    defaultPath("profileCardProperty"),
    deleteable(),
    updateable()
], _ProfileCardProperty);
export { _ProfileCardProperty };
export const ProfileCardProperty = graphInvokableFactory(_ProfileCardProperty);
/**
* Profilecard properties
*/
let _ProfileCardProperties = class _ProfileCardProperties extends _GraphCollection {
};
_ProfileCardProperties = __decorate([
    defaultPath("profileCardProperties"),
    getById(ProfileCardProperty),
    addable()
], _ProfileCardProperties);
export { _ProfileCardProperties };
export const ProfileCardProperties = graphInvokableFactory(_ProfileCardProperties);
