// @fluentui/react-migration-v8-v9/lib/components/Stack/StackShim.styles.js
import { __styles } from '@griffel/react';
export const useStackStyles = /*#__PURE__*/__styles({
  root: {
    mc9l5x: "f22iagw",
    Beiy3e4: "f1vx9l62",
    Eh141a: "flvyvdh",
    a9b677: "f14z66ap",
    Bqenvij: "f11ysow2",
    B7ck84d: "f1ewtqcl",
    Hc4np9: "f4sjk8j",
    Btkizpb: "f1d54kgm"
  },
  horizontal: {
    Beiy3e4: "f1063pyq"
  },
  verticalFill: {
    Bqenvij: "f1l02sjl"
  },
  reversedVertical: {
    Beiy3e4: "f1gkdon0"
  },
  reversedHorizontal: {
    Beiy3e4: "f1oztnx0"
  },
  disableShrink: {
    Btkizpb: "f1605gwb"
  },
  wrap: {
    mc9l5x: "f22iagw",
    Eh141a: "fni485r",
    Bqenvij: "f1l02sjl"
  },
  inner: {
    mc9l5x: "f22iagw",
    Beiy3e4: "f1vx9l62",
    Eh141a: "fni485r",
    B7ck84d: "f1ewtqcl",
    B2u0y6b: "f1w39fev",
    Btkizpb: "f1d54kgm"
  },
  innerWidth: {
    a9b677: "fly5x3f"
  }
}, {
  d: [".f22iagw{display:flex;}", ".f1vx9l62{flex-direction:column;}", ".flvyvdh{flex-wrap:nowrap;}", ".f14z66ap{width:auto;}", ".f11ysow2{height:auto;}", ".f1ewtqcl{box-sizing:border-box;}", ".f4sjk8j>*{text-overflow:ellipsis;}", ".f1d54kgm>*:not(.ms-StackItem){flex-shrink:1;}", ".f1063pyq{flex-direction:row;}", ".f1l02sjl{height:100%;}", ".f1gkdon0{flex-direction:column-reverse;}", ".f1oztnx0{flex-direction:row-reverse;}", ".f1605gwb>*:not(.ms-StackItem){flex-shrink:0;}", ".fni485r{flex-wrap:wrap;}", ".f1w39fev{max-width:100vw;}", ".fly5x3f{width:100%;}"]
});
export const useFlexGrowStyles = /*#__PURE__*/__styles({
  "1": {
    Bh6795r: "fqerorx"
  },
  "2": {
    Bh6795r: "fv68sux"
  },
  "3": {
    Bh6795r: "f1g5s877"
  },
  inherit: {
    Bh6795r: "fbrulcr"
  },
  initial: {
    Bh6795r: "fqq5sy3"
  },
  revert: {
    Bh6795r: "f1kc8nlq"
  },
  unset: {
    Bh6795r: "fepjonw"
  }
}, {
  d: [".fqerorx{flex-grow:1;}", ".fv68sux{flex-grow:2;}", ".f1g5s877{flex-grow:3;}", ".fbrulcr{flex-grow:inherit;}", ".fqq5sy3{flex-grow:initial;}", ".f1kc8nlq{flex-grow:revert;}", ".fepjonw{flex-grow:unset;}"]
});
export const useFlexAlignItemsStyles = /*#__PURE__*/__styles({
  baseline: {
    Bt984gj: "f1guyajx"
  },
  center: {
    Bt984gj: "f122n59"
  },
  start: {
    Bt984gj: "f6jr5hl"
  },
  end: {
    Bt984gj: "fgs5rwf"
  },
  stretch: {
    Bt984gj: "f1q9h2pe"
  },
  "space-between": {
    Bt984gj: "f1jkqsad"
  },
  "space-around": {
    Bt984gj: "ftnnmgl"
  },
  "space-evenly": {
    Bt984gj: "f1uhtmqw"
  }
}, {
  d: [".f1guyajx{align-items:baseline;}", ".f122n59{align-items:center;}", ".f6jr5hl{align-items:flex-start;}", ".fgs5rwf{align-items:flex-end;}", ".f1q9h2pe{align-items:stretch;}", ".f1jkqsad{align-items:space-between;}", ".ftnnmgl{align-items:space-around;}", ".f1uhtmqw{align-items:space-evenly;}"]
});
export const useFlexJustifyContentStyles = /*#__PURE__*/__styles({
  baseline: {
    Brf1p80: "f1uf0ky1"
  },
  center: {
    Brf1p80: "f4d9j23"
  },
  start: {
    Brf1p80: "fbhxue7"
  },
  end: {
    Brf1p80: "f9c4gz4"
  },
  stretch: {
    Brf1p80: "fsxf2b5"
  },
  "space-between": {
    Brf1p80: "f1869bpl"
  },
  "space-around": {
    Brf1p80: "f1iw3tyt"
  },
  "space-evenly": {
    Brf1p80: "figf6al"
  }
}, {
  d: [".f1uf0ky1{justify-content:baseline;}", ".f4d9j23{justify-content:center;}", ".fbhxue7{justify-content:flex-start;}", ".f9c4gz4{justify-content:flex-end;}", ".fsxf2b5{justify-content:stretch;}", ".f1869bpl{justify-content:space-between;}", ".f1iw3tyt{justify-content:space-around;}", ".figf6al{justify-content:space-evenly;}"]
});
