// @microsoft/sp-application-base/lib-commonjs/common/ApplicationId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SPScribeApplication = exports.SPScoreApplication = exports.SPSearchApplication = exports.SPSiteAdminApplication = exports.SPTopicLiteManagementApplication = exports.PortalLaunchPageApplication = exports.WorkbenchApplication = exports.VivaConnectionsAnalyticsApplication = exports.SPListHostApplication = exports.SPListFormApplication = exports.SPSystemPagesApplication = exports.CctrainingApplication = exports.ClassicSearchAnalyticsApplication = exports.SPAppStoreApplication = exports.SPTenantAppCatalogApplication = exports.MeeBridgeApplication = exports.SecureBrokerApplication = exports.SPBrokerLoginApplication = exports.VivaAmplifyApplication = exports.VivaHomeApplication = exports.IsolatedWebPartApplication = exports.ListApplication = exports.SPStartApplication = exports.SPHomeApplication = exports.SiteApplication = void 0;
/**
 * The sp-pages application component GUID (b6917cb1-93a0-4b97-a84d-7cf49975d4ec).
 */
exports.SiteApplication = 'b6917cb1-93a0-4b97-a84d-7cf49975d4ec';
/**
 * The sp-home application component GUID (1f019ae1-2de1-4f44-b723-00a6ec1d7445).
 */
exports.SPHomeApplication = '1f019ae1-2de1-4f44-b723-00a6ec1d7445';
/**
 * The sp-start-application component GUID (5b072552-f68a-4ac0-9a96-b2aa003f890c).
 */
exports.SPStartApplication = '5b072552-f68a-4ac0-9a96-b2aa003f890c';
/**
 * The listview-spfx-host application component GUID (b1ab4aaa-f779-405c-8683-d3a750b5d18d).
 */
exports.ListApplication = 'b1ab4aaa-f779-405c-8683-d3a750b5d18d';
/**
 * The isolated-web-part application component GUID (682c855d-1929-4602-a2fe-59962ccaea1a).
 */
exports.IsolatedWebPartApplication = '682c855d-1929-4602-a2fe-59962ccaea1a';
/**
 * The viva-home application component GUID.
 */
exports.VivaHomeApplication = 'eb765a5a-9fec-4d59-8cca-6c623b450f0a';
/**
 * The viva-amplify application component GUID.
 */
exports.VivaAmplifyApplication = '42dde860-0f24-4f29-a2c5-69fa26a70c27';
/**
 * The sp-broker-login application component GUID.
 */
exports.SPBrokerLoginApplication = 'efd6c8a7-f166-458c-905f-720088538e16';
/**
 * The sp-secure-broker application component GUID.
 */
exports.SecureBrokerApplication = 'ad084475-d3a1-478e-859b-7d9a31e86828';
/**
 * The mee-bridge-application component GUID.
 */
exports.MeeBridgeApplication = 'c22967c0-666a-4f03-a8c4-545efb9183e1';
/**
 * The sp-tenant-app-catalog-application component GUID.
 */
exports.SPTenantAppCatalogApplication = 'd0b6bbe9-3ee2-456a-8747-c9360f97a4bf';
/**
 * The sp-app-store-application component GUID.
 */
exports.SPAppStoreApplication = '854dd389-7596-43cb-91ad-75d41e24887a';
/**
 * The classic-search-analytics-application component GUID.
 */
exports.ClassicSearchAnalyticsApplication = '0cc4780a-78c5-4c54-9a83-20b4923d1a7a';
/**
 * The content-training-application component GUID.
 */
exports.CctrainingApplication = 'fca88529-7990-4631-aa5c-748ffe04799b';
/**
 * The sp-system-page-application component GUID.
 */
exports.SPSystemPagesApplication = '7497f5b1-726b-46d8-9798-6be6bbef75c5';
/**
 * The sp-list-form-application component GUID.
 */
exports.SPListFormApplication = '5dd2645c-2459-4b4d-a98a-89293a82323f';
/**
 * The sp-list-host-application component GUID.
 */
exports.SPListHostApplication = '9f012051-e030-4487-9eb7-3365d03be182';
/**
 * The viva-connection-analytics-application component GUID.
 */
exports.VivaConnectionsAnalyticsApplication = 'ef31f967-a8ff-477b-86d9-2a0c71276ef';
/**
 * The workbench-application component GUID.
 */
exports.WorkbenchApplication = '8be81a5c-af38-4bb2-af97-afa3b64dfbed';
/**
 * The sp-portal-laucnh-application component GUID.
 */
exports.PortalLaunchPageApplication = 'e51f2740-1bd0-4a54-83f7-e6d419283823';
/**
 * The sp-topic-lite-management-application component GUID.
 */
exports.SPTopicLiteManagementApplication = 'c3af8226-8f2d-4106-beb1-3c5cd7a0f663';
/**
 * The sp-site-admin-application component GUID.
 */
exports.SPSiteAdminApplication = '97e1e25f-5005-4973-9ff3-885603aebfff';
/**
 * The sp-search-application component GUID.
 */
exports.SPSearchApplication = 'c1c115a8-896b-4a53-a73c-cde6b8ca00f8';
/**
 * The sp-score-application component GUID.
 */
exports.SPScoreApplication = '579de7ba-edb6-454c-9d12-b7311379e662';
/**
 * The sp-scribe-application component GUID.
 */
exports.SPScribeApplication = '806f1922-71fc-44a2-a6cc-2b8ffdfec5fb';
