// @pnp/graph/mail/categories.js
import { __decorate } from "tslib";
import { _GraphCollection, _GraphInstance, graphInvokableFactory } from "../graphqueryable.js";
import { addable, defaultPath, deleteable, getById, updateable } from "../decorators.js";
// import { graphPost } from "@pnp/graph";
// import { body } from "@pnp/queryable";
/**
 * Outlook
 */
export class _Outlook extends _GraphInstance {
    get masterCategories() {
        return MasterCategories(this);
    }
}
export const Outlook = graphInvokableFactory(_Outlook);
/**
 * Describes an Outlook Category instance
 */
let _OutlookCategory = class _OutlookCategory extends _GraphInstance {
};
_OutlookCategory = __decorate([
    deleteable(),
    updateable()
], _OutlookCategory);
export { _OutlookCategory };
export const OutlookCategory = graphInvokableFactory(_OutlookCategory);
/**
 * Categories
 */
let _MasterCategories = class _MasterCategories extends _GraphCollection {
};
_MasterCategories = __decorate([
    defaultPath("masterCategories"),
    getById(OutlookCategory),
    addable()
], _MasterCategories);
export { _MasterCategories };
export const MasterCategories = graphInvokableFactory(_MasterCategories);
