// @fluentui/react-migration-v8-v9/lib-commonjs/components/Menu/shimMenuProps.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    shimMenuHeaderProps: function() {
        return shimMenuHeaderProps;
    },
    shimMenuItemCheckboxProps: function() {
        return shimMenuItemCheckboxProps;
    },
    shimMenuItemProps: function() {
        return shimMenuItemProps;
    },
    shimMenuProps: function() {
        return shimMenuProps;
    }
});
const _interop_require_wildcard = require("@swc/helpers/_/_interop_require_wildcard");
const _react = /*#__PURE__*/ _interop_require_wildcard._(require("react"));
const _react1 = require("@fluentui/react");
const shimMenuProps = (props)=>{
    return {
        hasIcons: props === null || props === void 0 ? void 0 : props.items.some((i)=>i.iconProps),
        hasCheckmarks: props === null || props === void 0 ? void 0 : props.items.some((i)=>i.canCheck || i.checked)
    };
};
const shimMenuItemProps = (props)=>{
    //TODO: Handle menuItem.onRenderIcon
    const icon = props.iconProps ? /*#__PURE__*/ _react.createElement(_react1.Icon, props.iconProps) : undefined;
    return {
        ...props,
        'aria-label': props.ariaLabel,
        children: props.text,
        icon,
        secondaryContent: props.secondaryText
    };
};
const shimMenuItemCheckboxProps = (props)=>{
    return {
        ...shimMenuItemProps(props),
        name: props.text || 'name',
        value: props.value || 'value'
    };
};
const shimMenuHeaderProps = (props)=>{
    var _props_sectionProps;
    return {
        children: (_props_sectionProps = props.sectionProps) === null || _props_sectionProps === void 0 ? void 0 : _props_sectionProps.title
    };
};
